clients MUST send Finished when using 0-RTT

kazuho opened the pull request:
since servers are allowed to postpone processing of the application data sent in 0-RTT until seeing client's Finished.

This is what some deployments already do and what others are interested in doing[1].

[1] https://lists.w3.org/Archives/Public/ietf-http-wg/2017AprJun/0124.html

davegarrett commented on draft-ietf-tls-tls13.md:
```diff
@@ -2939,6 +2939,12 @@ servers MUST process the client's ClientHello and then immediately
+Servers are permitted to buffer and postpone the processing of application
+data sent in 0-RTT until they receive client's Finished. Therefore, when
+using 0-RTT, clients MUST proceed the handshake to the point of sending
+their Finished messages even if there are no application data to be sent
```
nit: Has plural for clients, but singular for handshake, thus Finished messages should be singular, nor plural, for consistency. (one handshake has one Finished message from the client) So, instead of "their Finished messages", do "the Finished message". (also "the" instead of "their to match use in prior paragraph)

kazuho replied:
Thank you for pointing that out! Fixed in ea9ad23

kaduk commented:
If we're going to provide a justification for requiring a Finished, that's not the thing I'd put at the top of the list -- the Finished message is needed to get the anti-tampering property of the handshake!

martinthomson commented on draft-ietf-tls-tls13.md:
```diff
@@ -2939,6 +2939,12 @@ servers MUST process the client's ClientHello and then immediately
 EndOfEarlyData message.
 
+Servers are permitted to buffer and postpone the processing of application
+data sent in 0-RTT until they receive client's Finished. Therefore, when
```
It might help to at least hint at why this strategy might be useful.  That is, if this treatment is uniformly applied by a server, the data to which it is applied cannot ever be replayed, even though TLS does not provide replay protection.

martinthomson commented on draft-ietf-tls-tls13.md:
```diff
@@ -2939,6 +2939,12 @@ servers MUST process the client's ClientHello and then immediately
+data sent in 0-RTT until they receive client's Finished. Therefore, when
+using 0-RTT, clients MUST proceed the handshake to the point of sending
+the Finished message even if there are no application data to be sent
+in 1-RTT.
```
You are missing the necessary preconditions here: the requirement can only apply after the client has received the entire flight of server messages (ServerHello to server Finished).

martinthomson replied:
You are missing the necessary preconditions here: the requirement can only apply after the client has received the entire flight of server messages (ServerHello to server Finished).

kazuho commented:
@martinthomson Thank you for the suggestions. I have addressed the issues in 8d5e56e.

@kaduk Thank you for the comment. The intention is not to state the purpose of ClientFinished, but to state that the message can be used as a confirmation by the server that the 0-RTT data has not been replayed. I think I have clarified it in the commit above. Would you mind taking a look at it?

davegarrett commented on draft-ietf-tls-tls13.md:
```diff
@@ -2939,6 +2939,14 @@ servers MUST process the client's ClientHello and then immediately
+When using 0-RTT, clients MUST send its Finished message in response to
+the handshake messages sent from the server, even if there is no
+application data to be sent in 1-RTT. Servers MAY accept 0-RTT data but
+can still decide to buffer and postpone the processing of the payload
```
"but can" -> "yet"

davegarrett commented on draft-ietf-tls-tls13.md:
```diff
@@ -2939,6 +2939,14 @@ servers MUST process the client's ClientHello and then immediately
+the handshake messages sent from the server, even if there is no
+application data to be sent in 1-RTT. Servers MAY accept 0-RTT data but
+can still decide to buffer and postpone the processing of the payload
+until the client sends Finished, which can be used as a confirmation
```
drop "a"

davegarrett commented on draft-ietf-tls-tls13.md:
```diff
@@ -2939,6 +2939,14 @@ servers MUST process the client's ClientHello and then immediately
+application data to be sent in 1-RTT. Servers MAY accept 0-RTT data but
+can still decide to buffer and postpone the processing of the payload
+until the client sends Finished, which can be used as a confirmation
+that the data was not replayed. Using such strategy is beneficial when
```
"such strategy" -> "such a strategy"

davegarrett commented on draft-ietf-tls-tls13.md:
```diff
@@ -2939,6 +2939,14 @@ servers MUST process the client's ClientHello and then immediately
+can still decide to buffer and postpone the processing of the payload
+until the client sends Finished, which can be used as a confirmation
+that the data was not replayed. Using such strategy is beneficial when
+tolerance against replay attacks depend on the payload.
```
"depend" -> "depends"

davegarrett commented on draft-ietf-tls-tls13.md:
```diff
@@ -2939,6 +2939,14 @@ servers MUST process the client's ClientHello and then immediately
 send the ServerHello, rather than waiting for the client's
 EndOfEarlyData message.
 
+When using 0-RTT, clients MUST send its Finished message in response to
```
"clients" -> "a client"
("clients" is plural, yet "its" & "message" are singular; should stay consistent)

kazuho commented:
@davegarrett Thank you for the suggestions! Applied.

davegarrett commented on draft-ietf-tls-tls13.md:
```diff
@@ -2939,6 +2939,14 @@ servers MUST process the client's ClientHello and then immediately
 send the ServerHello, rather than waiting for the client's
 EndOfEarlyData message.
 
+When using 0-RTT, a client MUST send its Finished message in response to
```
@kaduk Dropping the "When using 0-RTT, " from the start might help alleviate your concern. That gets rid of that qualification. Optionally also add "always", e.g. "A client MUST always send its Finished..."

ekr commented:
I think Kaduk's suggestion is the right one here. See PR#1018

kazuho commented:
Thank you for all! I am glad to see the provision get into the specification.
